### PR TITLE
adds count num sequences and tokens metric

### PR DESCRIPTION
**Pull request by mosheraboh**

### CODE REVIEW

File: fuse/eval/metrics/sequence_gen/metrics_seq_gen_common.py [hunk 1]
```diff
@@ -18,13 +18,99 @@
...
 from functools import partial
-
+from copy import copy
 import torch
 import numpy as np
 
 from fuse.eval.metrics.metrics_common import MetricPerBatchDefault
 
 
+class MetricCountSeqAndTokens(MetricPerBatchDefault):
```
Comment (SagiPolaczek) [inline fuse/eval/metrics/sequence_gen/metrics_seq_gen_common.py:28]: General question:
I'm not sure counting the sequences and tokens should be defined as metric. I don't have another suggestion it's just sounds weird :) 

What do you think of that?

File: fuse/eval/metrics/sequence_gen/metrics_seq_gen_common.py [hunk 2]
```diff
@@ -18,13 +18,99 @@
...
+    def __init__(
+        self,
+        encoder_input: str,
+        ignore_index: Optional[int] = None,
+        state: Optional[dict] = None,
+        **kwargs: dict,
+    ) -> None:
+        """
+        :param encoder_input: key to the encoder_input
+        :param ignore_index: token_id to ignore (not to count), typically pad token id
```
Comment (SagiPolaczek) [inline fuse/eval/metrics/sequence_gen/metrics_seq_gen_common.py:42]: I think it should be able to support a list of token ids to ignore. Unless you want to enforce the user to ignore only the PAD one.

File: fuse/eval/metrics/sequence_gen/metrics_seq_gen_common.py [hunk 3]
```diff
@@ -18,13 +18,99 @@
...
+        :param state: the sequence count and token count to continue for. Should be restored when we continue training.
+                    use get_state() to get the state and save it upon checkpointing,
+        :param kwargs: additional super class arguments
+        """
+        super().__init__(
+            seq_num="seq_num",  # collect log_probs - output of  _count_seq_and_tokens_update
```
Comment (SagiPolaczek) [inline fuse/eval/metrics/sequence_gen/metrics_seq_gen_common.py:48]: obsolete comments in this line and the following one

File: fuse/eval/metrics/sequence_gen/metrics_seq_gen_common.py [hunk 4]
```diff
@@ -18,13 +18,99 @@
...
+        self._state["token_num"] += token_num_total
+        return copy(self._state)
+
+    def get_state(self) -> dict:
+        return copy(self._state)
+
+
+def _count_seq_and_tokens_update(
+    batch_dict: dict,
+    encoder_input_key: str,
```
Comment (SagiPolaczek) [inline fuse/eval/metrics/sequence_gen/metrics_seq_gen_common.py:84]:
```python 
encoder_input_key: Union[str, None] 
```
or 
```python
encoder_input_key: Optional[str]
```

File: fuse/eval/metrics/sequence_gen/metrics_seq_gen_common.py [hunk 5]
```diff
@@ -18,13 +18,99 @@
...
+        if state is None:
+            self._state = {"seq_num": 0, "token_num": 0}
+        else:
+            assert "seq_num" in state
+            assert "token_num" in state
+            self._state = state
+
+    def _count_seq_and_tokens_compute(
+        self,
+        seq_num: List[np.ndarray],
```
Comment (SagiPolaczek) [inline fuse/eval/metrics/sequence_gen/metrics_seq_gen_common.py:68]: `seq_num` will be a numpy array such that each entry represents a batch? If so, how often the metrics being calculate? each epoch? 

I forgot these :)

File: fuse/eval/metrics/sequence_gen/metrics_seq_gen_common.py [hunk 6]
```diff
@@ -18,13 +18,99 @@
...
+        token_num: List[np.ndarray],
+    ) -> float:
```
Comment (SagiPolaczek) [inline fuse/eval/metrics/sequence_gen/metrics_seq_gen_common.py:70]: returns a dict

File: fuse/eval/metrics/sequence_gen/metrics_seq_gen_common.py [hunk 7]
```diff
@@ -18,13 +18,99 @@
...
+
+    def get_state(self) -> dict:
+        return copy(self._state)
+
+
+def _count_seq_and_tokens_update(
+    batch_dict: dict,
+    encoder_input_key: str,
+    ignore_index: Optional[int] = None,
+) -> Tuple[torch.Tensor, torch.Tensor]:
```
Comment (SagiPolaczek) [inline fuse/eval/metrics/sequence_gen/metrics_seq_gen_common.py:86]:
```python
-> dict[str, Tensor]
```

**Review by SagiPolaczek — DISMISSED** [review]
LGTM, Thanks!

I added questions + typing mistakes + artifacts inline. 
Doesn't change the logic.

File: fuse/eval/metrics/sequence_gen/metrics_seq_gen_common.py (hunk 1 above)
Comment (mosheraboh) [reply fuse/eval/metrics/sequence_gen/metrics_seq_gen_common.py:28]: It uses the metric mechanism, and it's ok to me that it just counts some stats.

File: fuse/eval/metrics/sequence_gen/metrics_seq_gen_common.py (hunk 2 above)
Comment (mosheraboh) [reply fuse/eval/metrics/sequence_gen/metrics_seq_gen_common.py:42]: I went with just one to be more efficient - typically we would like to just skip the padding.

File: fuse/eval/metrics/sequence_gen/metrics_seq_gen_common.py (hunk 3 above)
Comment (mosheraboh) [reply fuse/eval/metrics/sequence_gen/metrics_seq_gen_common.py:48]: 👍

File: fuse/eval/metrics/sequence_gen/metrics_seq_gen_common.py (hunk 5 above)
Comment (mosheraboh) [reply fuse/eval/metrics/sequence_gen/metrics_seq_gen_common.py:68]: each sub epoch and each entry is a batch.

File: fuse/eval/metrics/sequence_gen/metrics_seq_gen_common.py (hunk 4 above)
Comment (mosheraboh) [reply fuse/eval/metrics/sequence_gen/metrics_seq_gen_common.py:84]: It's. a must. Why optional?

File: fuse/eval/metrics/sequence_gen/metrics_seq_gen_common.py (hunk 7 above)
Comment (mosheraboh) [reply fuse/eval/metrics/sequence_gen/metrics_seq_gen_common.py:86]: 👍

File: fuse/eval/metrics/sequence_gen/metrics_seq_gen_common.py (hunk 6 above)
Comment (mosheraboh) [reply fuse/eval/metrics/sequence_gen/metrics_seq_gen_common.py:70]: 👍

**Review by mosheraboh — COMMENTED** [review]
Thanks for the review and useful comments @SagiPolaczek

**Review by SagiPolaczek — APPROVED** [review]
LGTM

**Comment by SagiPolaczek** [comment]
Merging it to match inner-source code.
